src/vit-testing/mainnet-lib/src/db_sync/in_memory.rs

```diff
@@ -0,0 +1,353 @@
+    ///
+    /// # Panics
+    ///
+    /// On accessing db sync state
+    pub fn connect_to_node(self, node: &mut InMemoryNode) -> SharedInMemoryDbSync {
+        let mut observer = block_on(async {
+            node.observe(Channel::Bounded(1).into())
+                .await
+                .expect("observer")
+        });
+
+        let shared_db_sync = Arc::new(RwLock::new(self));
+        let db_sync = shared_db_sync.clone();
+
+        let handle = tokio::spawn(async move {
+            loop {
+                let block = observer.next().await;
+                if let Some(block) = block {
+                    db_sync.write().unwrap().on_block_propagation(&block);
+                }
+            }
+        });
+
+        SharedInMemoryDbSync {
+            update_thread: handle,
+            db_sync: shared_db_sync,
+        }
+    }
+
+    /// Retrieves db sync content as string
+    ///
+    /// # Errors
+    ///
+    /// On deserialization issues
+    pub fn try_as_string(&self) -> Result<String, serde_json::Error> {
+        serde_json::to_string_pretty(&self)
+    }
+
+    /// Accept new block
+    ///
+    /// # Panics
+    ///
+    /// On integer overflow
+    pub fn on_block_propagation(&mut self, block: &Block) {
+        self.blocks.push(block.clone());
+
+        let mut transactions = vec![];
+        let bodies = block.transaction_bodies();
+        for i in 0..bodies.len() {
+            let outputs = bodies.get(i).outputs();
+
+            for i in 0..outputs.len() {
+                let output = outputs.get(i);
+                let stake = output.amount().coin();
+                self.stakes
+                    .entry(output.address().to_hex())
+                    .and_modify(|x| *x = x.checked_add(&stake).unwrap())
+                    .or_insert(stake);
+            }
+
+            transactions.push(Transaction::new(
+                &bodies.get(i),
+                &TransactionWitnessSet::new(),
+                block.auxiliary_data_set().get(u32::try_from(i).unwrap()),
+            ));
+        }
+
+        self.transactions
+            .insert(block.header().header_body().block_number(), transactions);
+    }
+
+    /// Query transaction by it's hash representation
+    #[must_use]
+    pub fn transaction_by_hash(&self, hash: &str) -> Vec<(&Block, &Transaction)> {
+        self.transactions
+            .iter()
+            .filter_map(|(block, txs)| {
+                if let Some(tx) = txs.iter().find(|tx| tx.to_hex() == hash) {
+                    let block = self
+                        .blocks
+                        .iter()
+                        .find(|x| x.header().header_body().block_number() == *block)
+                        .expect("data inconsitency. cannot find block with block no from tx");
+
+                    return Some((block, tx));
+                }
+                None
+            })
+            .collect()
+    }
```

Comment:
I might be wrong but how can you return `Some()` and `None ` if the return type is not an `Option`?
Also I would expect to have a return type with one transaction (if it exist) and one or many blocks, maybe something like `Some(HashMap<Transaction, Vec<Block>>)` ?
Also I would avoid abbreviations in the `.expect("Data inconsistency. Cannot find block with block number from the transaction") `

Reply:
`filter_map` is doing magic with mapping elements and in the same time filtering it. That's why we need to tell them if we want to filter out some element (by returning None) or how we want to map elements which are not filtered out (by returning Some(....)). During `collect` phase of iterator, it removes Option<...> wrapper so we don't need to do it on our own. I will remove this nasty expect. thanks!